### PR TITLE
New resource - `azurerm_active_directory_domain_service_trust`

### PR DESCRIPTION
**Pull request by magodo**
This PR adds a new resource `azurerm_active_directory_domain_service_trust`, which is actually a property of the Azure resource: Azure Active Directory Domain Service. This modeling is inspired by the [corresponding powershell cmdlet](https://docs.microsoft.com/en-us/azure/active-directory-domain-services/create-resource-forest-powershell#create-the-managed-domain-side-of-the-trust-relationship).

This resource represents a trust relationship between an AADDS and a on-premise ADDS. To setup the E2E testing, there are following steps needed:

- Setup a on-premise ADDS (manually)
- Setup an AADDS (can via Terraform)
- Setup the network for the on-premise ADDS and the AADDS (can via Terraform)
- Configure DNS in the on-premise ADDS (manually)
- Create an inbound trust in the on-premise ADDS (manually)

(See: https://docs.microsoft.com/en-us/azure/active-directory-domain-services/tutorial-create-forest-trust)

Therefore, in the acctest, I'm passing in all the needed information which are setup previously via env vars. Also the tests are run in sequential, as at most there is only one AADDS in each tenant.

## Test Result

```shell
❯ TF_ACC=1 go test -timeout=60m -v -run='TestAccDomainServiceTrust_' ./internal/services/domainservices
=== RUN   TestAccDomainServiceTrust_basic
--- PASS: TestAccDomainServiceTrust_basic (187.35s)
=== RUN   TestAccDomainServiceTrust_requiresImport
--- PASS: TestAccDomainServiceTrust_requiresImport (193.09s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/domainservices        380.445s
```

### CODE REVIEW

File: internal/services/domainservices/active_directory_domain_service_trust_resource.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+
+func (r DomainServiceTrustResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+		"domain_service_resource_id": {
```
Comment (manicminer) [inline internal/services/domainservices/active_directory_domain_service_trust_resource.go:39]: We can shorten this to `domain_service_id`

```suggestion
		"domain_service_id": {
```

File: internal/services/domainservices/resourceids.go [hunk 2]
```diff
@@ -2,3 +2,5 @@ package domainservices
 
 //go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=DomainService -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.AAD/domainServices/DomainService1/initialReplicaSetId/replicaSetID
 //go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=DomainServiceReplicaSet -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.AAD/domainServices/DomainService1/replicaSets/replicaSetID
+//go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=DomainServiceTrust -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.AAD/domainServices/DomainService1/trusts/trust1
+//go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=DomainServiceResource -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.AAD/domainServices/DomainService1
```
Comment (manicminer) [inline internal/services/domainservices/resourceids.go:6]: This resource is already defined with existing parsers/validators as `DomainService`. To avoid confusion, I think it would be better to parse that, removing the custom suffix as needed?

File: website/docs/r/active_directory_domain_service_trust.html.markdown [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  trusted_domain_dns_ips     = ["10.1.0.3", "10.1.0.4"]
+  password                   = "Password123"
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `domain_service_resource_id` - (Required) The `resource_id` of the Active Directory Domain Service. Changing this forces a new Active Directory Domain Service Trust to be created.
```
Comment (manicminer) [inline website/docs/r/active_directory_domain_service_trust.html.markdown:38]: Note to rename this as per the above comment

File: internal/services/domainservices/active_directory_domain_service_trust_resource_test.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		AADDSResourceGroupName: aaddsRgName,
+		AADDSName:              aaddsName,
+		ADDSFqdn:               addsFqdn,
+		ADDSIp1:                addsIp1,
+		ADDSIp2:                addsIp2,
+		TrustPassword:          password,
+	}, nil
+}
+
+func TestAccDomainServiceTrust_basic(t *testing.T) {
```
Comment (manicminer) [inline internal/services/domainservices/active_directory_domain_service_trust_resource_test.go:75]: In terms of concurrency, these are probably OK for now in their current form, but we do need to look at handling concurrency with AADDS in general since it's tested in a few places now.

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @magodo, thanks for this PR! I've made some minor comments below and I'm going to have a go at running these tests later today.

File: internal/services/domainservices/resourceids.go (hunk 2 above)
Comment (magodo) [reply internal/services/domainservices/resourceids.go:6]: The suffix (`trusts/trust1`) is needed as there can be multiple trusts configured for a AADDS, the `trust1` corresponds to the friendly name of the trust.

File: internal/services/domainservices/active_directory_domain_service_trust_resource_test.go (hunk 4 above)
Comment (magodo) [reply internal/services/domainservices/active_directory_domain_service_trust_resource_test.go:75]: What does the `a few places` mean? If they all belong to the acc test cases, then we can make all of them as sequential tests?

File: internal/services/domainservices/resourceids.go (hunk 2 above)
Comment (manicminer) [reply internal/services/domainservices/resourceids.go:6]: Sure, but I meant the DomainServiceResource ID represents the same thing as the DomainService ID

File: internal/services/domainservices/active_directory_domain_service_trust_resource_test.go (hunk 4 above)
Comment (manicminer) [reply internal/services/domainservices/active_directory_domain_service_trust_resource_test.go:75]: There are also tests in different service packages that depend on AADDS

Comment (magodo) [reply internal/services/domainservices/active_directory_domain_service_trust_resource_test.go:75]: Then should we also guarantee that they are running in sequential?

File: internal/services/domainservices/resourceids.go (hunk 2 above)
Comment (magodo) [reply internal/services/domainservices/resourceids.go:6]: Currently, `DomainServiceResource` is only used for logging.. I can remove it then. Whilst, I'm wondering why did we choose the long form `/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.AAD/domainServices/DomainService1/initialReplicaSetId/replicaSetID` to represent the AADDS, rather than the `DomainServiceResource` form at the first place? As the former is hard for users to identify (imagin when users want to import it), and also depends on the ordering of the replica sets.

File: internal/services/domainservices/active_directory_domain_service_trust_resource_test.go (hunk 4 above)
Comment (manicminer) [reply internal/services/domainservices/active_directory_domain_service_trust_resource_test.go:75]: That's not possible at the moment, as they are in different service packages and built/tested separately.

File: internal/services/domainservices/resourceids.go (hunk 2 above)
Comment (manicminer) [reply internal/services/domainservices/resourceids.go]: We use a contrived ID that includes the initial replica set in order to overcome a conflict of interest between Terraform and the API - we have to keep track of the replica set that is implicitly created with the resource, so that we can also manage additional replica sets.

Comment (magodo) [reply internal/services/domainservices/resourceids.go]: It has been removed now.

File: internal/services/domainservices/active_directory_domain_service_trust_resource_test.go (hunk 4 above)
Comment (magodo) [reply internal/services/domainservices/active_directory_domain_service_trust_resource_test.go:75]: I see.

**Comment by magodo** [comment]
@manicminer Is there anything I can do to make this PR merge?

**Review by manicminer — APPROVED** [review]
@magodo Thanks for updating. We don't have a way to easily test at the moment but this LGTM and it's passing for you. Thanks again 👍
